### PR TITLE
win, process: EINVAL for invalid uv_kill signum

### PR DESCRIPTION
**Pull request by bzoz**
Changes the way invalid signals are handled by `uv_kill`. If the `signum` is invalid - below `0` or `NSIG` or above - `UV_EINVAL` will be returned even if target process does not exists. This makes `uv_kill` on Windows behave like on Linux systems.

Right now `uv_kill` will return `UV_ESRCH` if the process does not exists or `UV_ENOSYS` for any `signum` other than `0`, `SIGTERM`, `SIGKILL` or `SIGINT`. 

Refs: https://github.com/nodejs/node/pull/16944

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/590/

### CODE REVIEW

**Comment by bzoz** [comment]
It looks like [there are issues with NSIG on FreeBSD](https://bugs.python.org/issue20584), making the test fail on that platform

**Comment by santigimeno** [comment]
What about using [`SIGRTMAX`](https://stackoverflow.com/a/32427903/1562178). It looks like `NSIG` should be `SIGRTMAX + 1`?

**Comment by bzoz** [comment]
Not available on Windows

**Comment by santigimeno** [comment]
And doing something like this?

```
#ifdef SIGRTMAX
#define UV_NSIG SIGRTMAX + 1
#else
#define UV_NSIG NSIG
#endif

TEST_IMPL(kill_invalid_signum) {
  ASSERT(uv_kill(0, -1) == UV_EINVAL);
  ASSERT(uv_kill(0, UV_NSIG) == UV_EINVAL);
  ASSERT(uv_kill(0, 1024) == UV_EINVAL);
  MAKE_VALGRIND_HAPPY();
  return 0;
}
```

**Comment by bzoz** [comment]
That looks like it will work, I'll give it a spin.

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/597/

**Comment by bzoz** [comment]
Now [aix fails](https://ci.nodejs.org/job/libuv-test-commit-aix/600/nodes=aix61-ppc64/console) and [zos does not compile](https://ci.nodejs.org/job/libuv-test-commit-zos/274/nodes=zos-s390x/console). It looks like it also [kills the test runner on freebsd](https://ci.nodejs.org/job/libuv-test-commit-freebsd/567/nodes=freebsd10-64/console).

**Comment by bzoz** [comment]
`NSIG` test only on Windows, for other platforms we will test `4096` as the "too high" signum. Now it will also find an unused pid.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/598/

File: test/test-signal.c [hunk 1]
```diff
@@ -22,6 +22,28 @@
...
 #include "task.h"
 
+#define MAX_PID 32768
+TEST_IMPL(kill_invalid_signum) {
+  int pid;
+
+  /* Find unused pid. */
+  pid = 1;
+  while (pid < MAX_PID && uv_kill(pid, 0) != UV_ESRCH)
+    ++pid;
```
Comment (bnoordhuis) [inline test/test-signal.c]: This is racy as heck - the pid can get used while the test is in progress.  I'm okay with making it a Windows-only test if that is easier for you.

**Comment by jBarz** [comment]
Try rebase to get rid of the z/OS failure

**Comment by bzoz** [comment]
Turns out, that Linux also tests if the process exists before validating the signal. Its just that kill with pid 0 is special on Linux, and means “this process group”. I’ve changed the `uv_kill` to emulate this on Windows too.
CI is now green (module unrelated failure): https://ci.nodejs.org/view/libuv/job/libuv-test-commit/605/, PTAL.

**Comment by bzoz** [comment]
@bnoordhuis @refack @cjihrig  this changed since you approved this, can you PTAL?

File: test/test-signal.c [hunk 2]
```diff
@@ -22,6 +22,30 @@
...
 
+#ifndef _WIN32
+#include <unistd.h>
+#endif
+
+TEST_IMPL(kill_invalid_signum) {
+  int pid;
+
+#ifdef _WIN32
+  pid = GetCurrentProcessId();
```
Comment (cjihrig) [inline test/test-signal.c]: `uv_os_getpid()`?

**Comment by bzoz** [comment]
Updated, PTAL

**Comment by cjihrig** [comment]
Still LGTM

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/627/

**Comment by bzoz** [comment]
Landed in https://github.com/libuv/libuv/commit/890eedaf59cea75faaa6a14b4248a472dcadb831

**Pull request closed without merging**